pixeltable/env.py

```diff
@@ -1135,18 +1146,44 @@ def debug_str(self) -> str:
         )
 
     def update(self, recorded_at: datetime.datetime, limit: int, remaining: int, reset_at: datetime.datetime) -> None:
-        # we always update everything, even though responses may come back out-of-order: we can't use reset_at to
-        # determine order, because it doesn't increase monotonically (the reset duration shortens as output_tokens
-        # are freed up - going from max to actual)
+        if self.recorded_at > recorded_at:
+            # The current state is more up-to-date than the update
+            _logger.debug(
+                f'Ignoring out-of-date update for {self.resource}. Current recorded_at: '
+                f'{self.recorded_at.strftime(TIME_FORMAT)}, update: {recorded_at.strftime(TIME_FORMAT)}'
+            )
+            return
         self.recorded_at = recorded_at
         self.limit = limit
         self.remaining = remaining
-        reset_delta = reset_at - self.reset_at
         self.reset_at = reset_at
-        # TODO: remove
-        _logger.debug(
-            f'Update {self.resource} rate limit: rem={self.remaining} reset={self.reset_at.strftime(TIME_FORMAT)} '
-            f'reset_delta={reset_delta.total_seconds()} recorded_delta={(self.reset_at - recorded_at).total_seconds()}'
+
+    def estimated_time_until_balance(self, target_remaining: int) -> float | None:
+        """Estimate time in seconds until remaining resources reaches target_remaining.
+        Assumes linear replenishment of resources over time.
+        Returns None if unable to estimate.
+        """
+        if self.remaining >= target_remaining:
+            return 0
+        if self.recorded_at >= self.reset_at:
+            return 0
+        if self.limit < target_remaining:
+            return None
+
+        # Estimate resource refill rate based on the recorded state and timestamps. Assumes linear refill.
+        refill_rate = (self.limit - self.remaining) / (self.reset_at - self.recorded_at).total_seconds()
+        if refill_rate <= 0:
```

Comment:
Isn't this impossible since the numerator and denominator must both be positive if we reach this point?

Reply:
I think that is correct, replaced with an assertion